planning/MartenV4.md

```diff
@@ -0,0 +1,405 @@
+  easier to build. We can do this either by continuing to build Expressions and compiling to Funcs, or we could re-introduce code compilation via LamarCompiler (which was originally
+  built into Marten pre-V1)
+
+**The potential downside is that this doesn't work well at all with `internal` properties, so we'll have to watch that**
+
+
+## Partial Updates
+
+Use native postgresql partial JSON updates wherever possible. Let's do a perf test on that first though.
+
+## Event Sourcing
+
+There's an existing conversation on GitHub about the [event sourcing improvements for v4](https://github.com/JasperFx/marten/issues/1307).
+
+### Event Store Metadata
+
+See the earlier conversation about document metadata up above, and also the next section on schema generation for the event store.
+
+What if we allow for custom event wrapper (what's now `IEvent` and `Event<T>`) types, and event stream types? So you could use custom
+metadata information right on those types? And possibly even put some logic into event stream creation to capture stream
+metadata?
+
+If so, those types could be something like this:
+
+```
+    public enum StreamState
+    {
+        CreateNew,
+        Append,
+        AppendOrCreate,
+        Pending,
+        History
+    }
+    
+    // This really represents a segment of the underlying stream
+    public abstract class EventStreamBase<TKey, TEvent, TEventBase> where TEvent : IEvent<TEventBase>
+    {
+        public TKey Id { get; set; }
+
+        // There's a little bit of optimization for inline
+        // projections by doing this
+
+        // Other folks want different mechanics
+        public StreamState State { get; set; }
+
+        public IList<TEvent> Events { get; } = new List<TEvent>();
+
+
+        // Data in here while it's being appended, but moves up above when
+        // it's pushed into the database
+        public IList<TEventBase> Pending { get; } = new List<TEventBase>();
+    }
+
+    public interface IEvent<TEventBase>
+    {
+        Guid Id { get; set; }
+        int Version { get; set; }
+        long Sequence { get; set; }
+        TEventBase Data { get; set; }
+        DateTimeOffset Timestamp { get; set; }
+
+        // Additional, user-defined metadata properties
+        // like user id or correlation id or whatever
+
+    }
+
+    // This might be too heavyweight
+    public interface IMultitenantEvent<TEventBase>: IEvent<TEventBase>
+    {
+        string TenantId { get; set; }
+    }
+
+The custom stream and event types could be used to define segregations between different types of event logs within a single application as part
+of the partitioning/sharding/segmenting strategy
+
+
+
+The schema generation would conform to the shape of the base event and stream wrapper much the way that document storage does in the Doc Db features.
```

Comment:
I'm fine with the new events, however, our current interface IEvent is not much useful (at least from my experience). Using it in the projections is cumbersome, as you cannot use the same way for aggregation.
In my opinion, it's needed to allow applying the fields from metadata by convention to the event data (eg. Timestamp, Version etc.).
We'd need to also consider how having the Metadata would impact our Append, Aggregation, Projections api.

Reply:
I'm actually voting to let users define their own `IEvent` this time around for their metadata scheme, and to allow for a bit more variance on identity of the streams.

I did add some notes about the inline projections so you *could* use the `IEvent` or `Event<T>` correctly in v4

Reply:
Custom IEvent would sound much better for me than the generic class 👍 However the more we could do to reduce the need of using interfaces or base classes the better. That's why I suggested some convention based metadata mapping.

Reply:
This is going to need more time and thought